lua-api-crates/table-funcs/src/lib.rs

```diff
@@ -203,6 +232,34 @@ fn has_value<'lua>(_: &'lua Lua, (table, value): (Table<'lua>, LuaValue)) -> mlu
     Ok(false)
 }
 
+fn equal<'lua>(lua: &'lua Lua, (table1, table2): (Table<'lua>, Table<'lua>)) -> mlua::Result<bool> {
```

Comment:
I'd suggest something like:

```rust
fn lua_value_eq(value1: &LuaValue, value2: &LuaValue) -> bool {
   match (value1, value2) {
      (LuaValue::Table(a), LuaValue::Table(b)) => lua_table_eq(a, b),
      (a, b) => a.eq(b)
   }
}

fn lua_table_eq(table1: &Table, table2: &Table) -> bool {
   let mut table1_len = 0;
   for pair in table1.pairs::<LuaValue, LuaValue>() {
        match pair {
           Ok(key, value) => {
             table1_len += 1;
             match table2.get(key.clone()) {
                Ok(value2) => {
                    if !lua_value_eq(value, value2) {
                         return false;
                    }
                 }
                 Err(_) => return false,
           }
           Err(_) => return false,
        }
    }
    let table2_len = table2.pairs::<LuaValue, LuaValue>().count();
    table1_len == table2_len
}
```

Reply:
I have updated the code to something more like this, but I can't easily get around the `.clone()`s before `.pairs()`, since `equal` itself can't accept two references like `lua_table_eq` does.

If I change `lua_table_eq` to not use references, then I have to add something similar in `lua_value_eq` anyways. 

If you know a good way to get around this problem I can try to improve the code further.